### PR TITLE
Issue/308 inc warmup logic

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Cleaned up logic for methods `draws_pd` and `stan_variable` (#308)
and expose CmdStanMCMC attributes `_draws_warmup` and `_draws_sampling` as
properties `num_draws_warmup` and `num_draws_sampling`

Added unit tests.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by mitzimorris** [comment]
bug was because `draws_pd` method tries to cache the pd, to avoid making extra copies, but need to check whether or not cached object includes/excludes warmup draws correctly.

**Comment by mitzimorris** [comment]
hi @ahartikainen or @imadmali - this needs code review if you have any free cycles

File: cmdstanpy/stanfit.py [hunk 1]
```diff
@@ -708,16 +723,23 @@ def draws_pd(
...
+
+        if inc_warmup and not self._save_warmup:
+            self._logger.warning(
+                'draws from warmup iterations not available,'
+                ' must run sampler with "save_warmup=True".'
+            )
+
+        num_draws = self._draws_sampling
+        if inc_warmup and self._save_warmup:
+            num_draws += self._draws_warmup
```
Comment (ahartikainen) [inline cmdstanpy/stanfit.py:735]: Is `self._draws_warmup` always a number (I think it is, just making sure)

**Review by ahartikainen — APPROVED** [review]

File: cmdstanpy/stanfit.py (hunk 1 above)
Comment (mitzimorris) [reply cmdstanpy/stanfit.py:735]: yes, because at line 725 there's a call to `self._assemble_draws()`

**Pull request merged**